casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -43,6 +43,25 @@ object ProtoUtil {
       }
     }
 
+  // todo test this method, and replace everything in isInMainChain
+  def isInMainChain[F[_]: Monad](
+      dag: BlockDag,
+      candidate: BlockMessage,
+      targetBlockHash: BlockHash
+  ): F[Boolean] =
+    if (candidate.blockHash == targetBlockHash) {
+      true.pure[F]
+    } else {
+      dag.dataLookup.get(targetBlockHash) match {
+        case Some(targetBlockMeta) =>
+          targetBlockMeta.parents.headOption match {
+            case Some(mainParentHash) => isInMainChain[F](dag, candidate, mainParentHash)
+            case None                 => false.pure[F]
+          }
+        case None => false.pure[F]
+      }
+    }
+
```

Comment:
Should I remove the original isInMainChain, which get the data from BlockStore and only used in SafetyOracle?

Reply:
Done